### DISCUSSION TITLE
Last some mini-batch of NeighborLoader contain no edges.

### DISCUSSION THREAD

**Discussion by kam6ri** [pyg-team/pytorch_geometric#3331] [Q&A] [answered]
I have a question about the behavior of torch_geometric.loader.NeighborLoader.
I have created a custom InMemoryDataset by referring to the documentation, and trying to get mini-batch graph using NeighborLoader.
But some last mini-batch have no edges.
Why the last mini-batch of NeighborLoader contain no edges?

I thought it was an isolated node, but my graph did not contain any isolated nodes.

This is my implementation with output,

```
from torch_geometric import utils

dataset = CustomInMemoryDataset(...)

print(dataset.data)
->
Data(x=[103527, 300], edge_index=[2, 16910925], edge_attr=[16910925], y=[103527], train_masks=[103527], val_masks=[103527], test_masks=[103527])

print(utils.contains_isolated_nodes(dataset.data.edge_index))
->  
False 

loader = NeighborLoader(dataset.data, num_neighbors = [100] * 2, batch_size=64, input_nodes=data.train_masks)

for batch in loader:
    print(batch)

-> 
Data(x=[1393, 300], edge_index=[2, 110999], edge_attr=[110999], y=[1393], train_masks=[1393], val_masks=[1393], test_masks=[1393], batch_size=64)
    ...
Data(x=[64, 300], edge_index=[2, 0], edge_attr=[0], y=[64], train_masks=[64], val_masks=[64], test_masks=[64], batch_size=64)
# Is this isolated nodes ?
...
```

**Comment by wsad1** [comment, marked as answer]
Yeah the last few nodes might be isolated. `contains_isolated_nodes(dataset.data.edge_index)` will return `False` in that case. Try running `contains_isolated_nodes(dataset.data.edge_index, dataset.data.num_nodes)` this should return `True`.

**Reply by kam6ri** [reply]
@wsad1 Thanks :) `contains_isolated_nodes(dataset.data.edge_index, dataset.data.num_nodes)` return `True`.

In addition, I tried to remove isolated nodes from dataset by using `remove_isolated_nodes(dataset.data.edge_index, dataset.data.edge_attr, dataset.data.num_nodes)`, and it return some tensor objects including mask tensor.
Can I directly replace the original dataset to remove isolated nodes?
The following code does not seem to be reasonable.
```
data = dataset.data
data.edge_index, data.edge_attr, mask = remove_isolated_nodes(data.edge_index, data.edge_attr, num_nodes=data.num_nodes)
data.x = data.x[mask]
data.train_masks = data.train_masks[mask]
data.test_masks = data.test_masks[mask]
data.val_masks = data.val_masks[mask]
data.y = data.y[mask]
```
or
```
for batch in loader:
    if batch.num_edges == 0:
        continue
```

**Reply by rusty1s** [reply]
Note that in the second option, you only skip mini-batches if all its nodes are isolated. However, either way is fine, although you should also be able to process mini-batches that contain no edges without any problems.

**Reply by kam6ri** [reply]
Yes, indeed. Would it be possible to run the node classification may not contain edges in a GNN? (Sorry if this is a rudimentary question.)

**Reply by rusty1s** [reply]
```python
edge_index = torch.tensor([[], []], dtype=torch.long)
x = torch.randn(5, 16)

conv = SAGEConv(16, 32)
out = conv(x, edge_index)
```
should work. In this case, the `SAGEConv` reduces to a linear transformation, as no neighborhood information can be aggregated.
